### ISSUE TITLE
make CompendiumNG installer

### ISSUE THREAD

**Issue by maiklos** [CompendiumNG/CompendiumNG#28]
make CompendiumNG installer

try: http://izpack.codehaus.org/izpack-maven-plugin/

**Comment by SebastianBoell** [comment]
I, and several others run Compendium through a cloud service or from an USB stick that allows to start it under Linux and Windows respectivel from the same installation. Possibly the installer could provide the option to choose starter's for different OS simultaniously (for cloud installations / USB key installations). In such a case it would include Compendium.sh as well as Compendium.bat (and possibly also a starter for Mac). Don't know if this is feasable because of e.g. path issues and icons in startup menues.
If not we certailly should include instructions on how to run Compendium in such a setting in our Wiki.

**Comment by SebastianBoell** [comment]
I thought about this and a simple solution would be to offer a simple option as part of the installation:
- Regular (Windows, Linux, Mac) installation
- USB / cloud drive installation

Option 2 would then simply put all the starters into the according folder.

**Comment by maiklos** [comment]
first release for testing pushed to the repository into the izpack-installer branch commit: 8ca890b4cb41281514a8eefa4c3802fa5b027e4c

**Comment by maiklos** [comment]
testing on windows7 & windows XP & linux:
a) shortcut panel is empty on windows (it is displayed correctly on linux) (critical)
b) shortcuts (startup menu) not created  win & linux (major)
c) [FIXED] missing .compendiumng directory in the user home directory on win (critical)
d) [FIXED] missing demo map (CNG displays a message box about missing xml file) (minor)
e) some graphics should be added to some of the panels (minor)
f) [FIXED] there should be credits panel - to mention project contributors (former and current)

**Comment by SebastianBoell** [comment]
I updated the text for the installer and pushed back. Also I have added suggestions for further text in the wiki: https://github.com/CompendiumNG/CompendiumNG/wiki/Installer-Screens

Finally, I suggest to change the variable "installer_banner" used in installer_readme.htm to refer to 'installer_banner.png' rather than 'header2.png'.

**Comment by maiklos** [comment]
a) shortcut panel is empty on windows (it is displayed correctly on linux) (critical)
 version of izpack upgraded to 5.0.0-beta10, that should fix the problem [unfortunately still empty]
b) shortcuts (startup menu) not created win & linux (major)
 this should be fixed too [still no menu or desktop icon in XinXP; and no desktop in Debian]
c) [FIXED] missing .compendiumng directory in the user home directory on win (critical)
 this should work also [FIXED, directory is now created (WinXP)]
d) [FIXED] missing demo map (CNG displays a message box about missing xml file) (minor)
 CNG is not trying to load the default map. Should it stay like this ?
e) some graphics should be added to some of the panels (minor)
f) [FIXED] there should be credits panel - to mention project contributors (former and current)
yes

**Comment by maiklos** [comment]
commit 6e3d6fe introduced some significant changes so there is chance that something got broken in the process.
currently I am aware of following:
g) [FIXED] wrong (big) icon in one of toolbars
h) [FIXED] importer/exporter validates against compendium.dtd (in default location) - not sure what is the default location and if it works when user starts compendium from other locations
i) [FIXED] problem with build not working on windows should be fixed now [FIXED, could compile installer]
j) [FIXED] opening links from help menu currently doesn't work
k) [FIXED] after installation compendium start with window of size of few pixels in the left/top corner
l) [FIXED] welcome screen have only 2 butttons - something is missing
m) [FIXED] all toolbars are minimized

**Comment by SebastianBoell** [comment]
n) [FIXED] the .bat startup file for windows is not created correctly, some of the variables are not filled:

start /b javaw -Xmx512m -Dlog4j.configuration=file:${user.setting.dir}/log4j.properties com.compendium.ProjectCompendium %1 %2 %3 %4 %5 %6 %7 %8 %9

**Comment by SebastianBoell** [comment]
o) [FIXED] it seems that opening CNG after a new install works only once (at least in linux) at a subsequent start I get the following error message. I tried this without creating a project or creating a project, the result is the same:

`INFO [main](ProjectCompendiumFrame.java:6159) Error:The local Derby Administration database was unable to be opened/created...
Table/View 'PROJECT' already exists in Schema 'APP'.

Please contact Compendium support staff.`

**Comment by maiklos** [comment]
a) [PENDING] **waiting** for retest
b) [PENDING] **waiting** for retest
c) [FIXED]
d) [FIXED] default map is loaded from ProjectTemplates/DefaultData.xml (configurable by  dir.project.templates)
e) [PENDING] **not sure what is this**
f)  [FIXED] there is about box - almost same content as splash screen.
g) [FIXED] by d97df
h) [FIXED] by placing compendium.dtd into app. root folder (separate issue for .dtd location override #132)
i) [FIXED]
j) [FIXED] 2b987
k) seems [FIXED] by d7f10a2
l) [FIXED] d97df
m) [FIXED] probably caused by k)
n)  [FIXED]]
o) [FIXED] by d7f10a2
p) [PENDING] help file refers to images that are in other location now. **need to fix** references in html

.. hope will not eat whole alphabet on this issue ;-)

**Comment by SebastianBoell** [comment]
I like the split of settings, data and application a lot. I think you have done some great work there. Given this improved situation I wonder if we should already include a means for CNG to re-reading data from the data directory after a new installation. We now advertise that it is easily possible to re-install without loosing data, but then CNG does not display the data that is there already and the open dialog does not seem to work.
q) [FIXED] I have changed the wiki site for our list of changes to fit our established versioning scheme. The link in the installer thus needs to be updated to: https://github.com/CompendiumNG/CompendiumNG/wiki/List-of-Changes-in-2.1.1

**Comment by maiklos** [comment]
I believe that as long as you connect to the same admin db then you should be able to open you previous projects.

you should be also able to have 2 or more installations of CNG to access the same admindb an therefore same projects (not at the same time though)

is there there a use case that doesn't work ?
Should we test such scenario ?

**Comment by maiklos** [comment]
q) [FIXED] in f3678

**Comment by SebastianBoell** [comment]
To clarify the use case is that a user deinstalls an old version and then installs a new version. Subsequently the projects from the old version are no longer displayed in CNG. It is as if they are not there, however, I still see them in the data directory. I was just playing around with the feature suggestions (backed up) and then installed the new version this morning but the features project is no longer displayed. Not sure if the admindb was overwritten during installation?
It may have to do with the uninstaller which I notices also clears all files in '.compendiumng' but leaves the directory intact. I guess the uninstaller should leave the user setting files on disc or should remove the directory as well.

**Comment by maiklos** [comment]
in that case run the installer and make it create setting files for you. As long as you install into the same directory
it should create same configuration. or you can install into another dir and change .compendiumng/main.properties to point to data dir in old installation. I am convinced it should work.

**Comment by SebastianBoell** [comment]
I managed to reproduced this and I think I found the reason for this.

It seems that the latest installer creates the 'data' 'export' and 'backups' directory inside the application directory. As a result of this data is no longer read from the CompendiumNG-Data folder. Also the uninstaller now removes everything including the data.

2 ways to procede:
1) use a compendium-ng data folder in the home directory as in earlier versions of 2.1.1
2) use a single compendiumng folder, but within it separate an 'application' folder. When the uninstaller is used only the 'application' folder should be removed as not to delete data.

**Comment by SebastianBoell** [comment]
r) [FIXED] should we ensure that an icon is used for the CNG starters?
s) under 'Debian' no desktop icon is created
t) under 'debian' three starters are created rather than only one (may be omittable for now if it is Debian specific)

**Comment by SebastianBoell** [comment]
I set up an old winXP laptop which we can use for WinXP testing, so this should reduce turn around times for testing XP :)
a) still no panel displayed in installer
b) no starter or desktop icon created
n) [FIXED] this looks better now, but there still seems to be an issue that hinders start-up:

@echo off
cd E:\Program Files\CompendiumNG
start /b javaw -Xmx512m -Dlog4j.configuration=file:E:\Documents and Settings\Owner/log4j.properties -jar CompendiumNG-jar-with-dependencies.jar  %1 %2 %3 %4 %5 %6 %7 %8 %9

**Comment by maiklos** [comment]
> To clarify the use case is that a user deinstalls an old version and then installs a new version. Subsequently the  
>  projects from the old version are no longer displayed in CNG. It is as if they are not there, however, I still see 
>  them in the data directory. I was just playing around with the feature suggestions (backed up) and then installed 
>  the new version this morning but the features project is no longer displayed. Not sure if the admindb was 
>  overwritten during installation?

No it wasn't overwritten. admindb is only created on start-up when admin database doesn't exist in data location
also installer doesn't even care about admindb - it just creates and sets data directory for your installation.
admindb is reused (if it exist) or created (if it doesn't) on startup of the CNG application

> It may have to do with the uninstaller which I notices also clears all files in '.compendiumng' but leaves the directory intact. I guess the uninstaller should leave the user setting files on disc or should remove the directory as well.

uninstaller deletes all CNG files (except for .compendiumng dir). If you uninstall you want to stop using the application and even configuration files should be deleted. However I admit that that the uninstaller could be more sophisticated (i.e. ask what parts should be removed etc.) I will check if it has such feature.

besides of that upgrade or re-installation should not be done via uninstall/install. You should simply install again into the same directory. installer will ask if it should overwrite *.properties files and silently overwrites
jar, resources and startup scripts

**Comment by SebastianBoell** [comment]
For easy reference I include all open issues:

a,b) no panel and no desktop and start menu icons
e) [PENDING] but (minor)
n) [FIXED] .bat starter does not work correctly
p) [PENDING] - did not check images in help
r) [FIXED]
s) no desktop icon under debian (minor)
t) I now "only" get 2 starter entries (down from 3) for CNG in my menu, but 3 for the uninstaller (up from 1)

Re: t) can we name the Unistaller from the generic 'Uninstaller' to 'Uninstall CNG' so that users clearly see what it is for?

By the way, saving data by simply moving the 'data' directory into a new installation works like a charm and certainly will be welcomed by many users. This makes manual backing up and moving data around much, much easier!

**Comment by maiklos** [comment]
a,b) no shortcut panel displayed on windows XP to create shortcuts. Probably failed to raise priviledges.
and as a result also  no desktop icons are created by installer -> #140
e) [PENDING] but (minor) -> IMO fixed.
n) [FIXED] .bat starter does not work correctly
p) [PENDING] - did not check images in help -> fixed
r) [FIXED]
s) no desktop icon under debian (minor) -> #139
t) I now "only" get 2 starter entries (down from 3) for CNG in my menu, but 3 for the uninstaller (up from 1)
   t) is probably related to  a,b so -> #140

I want to close the mega-multi-ticket since we have fixed what could be fixed and for the rest we have dedicated tickets.
